pkg/batch/executor.go

```diff
@@ -0,0 +1,103 @@
+package batch
+
+import (
+	"context"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/logging"
+)
+
+type BatchFn func() (interface{}, error)
+
+type DelayFn func(dur time.Duration)
+
+type Batcher interface {
+	BatchFor(key string, dur time.Duration, fn BatchFn) (interface{}, error)
+}
+
+type nonBatchingExecutor struct {
+}
+
+func (n *nonBatchingExecutor) BatchFor(key string, dur time.Duration, fn BatchFn) (interface{}, error) {
+	return fn()
+}
+
+type response struct {
+	v   interface{}
+	err error
+}
+
+type request struct {
+	key              string
+	dur              time.Duration
+	fn               BatchFn
+	responseCallback chan *response
+}
+
+type Executor struct {
+	requests chan *request
+	execs    chan string
+	keys     map[string][]*request
+	Logger   logging.Logger
+	Delayer  DelayFn
+}
+
+func NopExecutor() *nonBatchingExecutor {
+	return &nonBatchingExecutor{}
+}
+
+func NewExecutor(logger logging.Logger) *Executor {
+	return &Executor{
+		requests: make(chan *request),
+		execs:    make(chan string),
```

Comment:
If you'll create a buffered channel (i.e. size more than 1), logic will still be consistent but you'll have less open go routines.